File: sorting & searching/C++/binary_search.cc
/**
 * @file binary_search.cc
 * @author Hongyu Ding
 * @brief This is the code of binary search
 * @version 1.0
 * @date 2020-10-17
 * 
 * @copyright Copyright (c) 2020
 * 
 */
#include <iostream>
#include <vector>

/**
 * @brief bin_search returns the index of target in the array a
 * if not found, returns -1
 * O(log n)
 * 
 * @param a a sorted array
 * @param target the value we want to find
 * @return int 
 */
int bin_search(std::vector<int> &a, int target) {
    int ans = -1;
    int l = 0, r = a.size(), mid;
    while (l < r) {
        mid = (l + r) / 2;
        if (a[mid] < target) {
            l = mid + 1;
        } else {
            r = mid;
        }
    }
    if (l != a.size() && a[l] == target) ans = l;
    return ans;
}

int main() {
    int n, k;
    std::vector<int> array;

    std::cin >> n;
    for (int i = 0; i < n; i++) {
        int val = 0;
        std::cin >> val;
        array.push_back(val);
    }
    std::cin >> k;

    std::cout << bin_search(array, k) << std::endl;    

    return 0;
}

File: sorting & searching/Java/BinarySearch.java
/**
 * @file BinarySearch.java
 * @author Weilan Tao
 * @brief This is the code of the java version binary search
 * @version 1.0
 * @date 2020-10-18 EST
 *
 * @copyright Copyright (c) 2020
 *
 */


import java.util.*;

public class BinarySearch {

    /**
     * Iterative Binary Search
     * @param array
     * @param k
     * @return the index of the target value k in the array
     */
    public static int binarySearch(int[] array, int k){

        int l=0;
        int r=array.length-1;
        int mid;
        while(array[l]<array[r]){
            mid=l+(r-l)/2; // prevent overload

            if (array[mid] == k) {
                r = mid;
            }
            else {
                l = mid + 1;
            }
        }
        return l;
    }

    /**
     * bubble sort
     * @param array
     * @param len
     * return sorted array
     */
    public static void bubbleSort(int[] array,int len){
        int temp;
        for(int i =0; i<len-1; i++){
            for (int j=0;j<len-i-1;j++){
                if(array[j]>array[j+1]) {
                    temp = array[j];
                    array[j] = array[j + 1];
                    array[j + 1] = temp;
                }
            }
        }
    }

    public static void main(String args[]){
        System.out.println("enter the length of an array");
        Scanner s1= new Scanner(System.in);
        int x=s1.nextInt();
        int[] array= new int[x];

        System.out.println("initialize the array");
        for(int i=0; i<x; i++){
            array[i]=s1.nextInt();
        }
        bubbleSort(array,x);
        System.out.println("enter the target value");
        int k=s1.nextInt();
        
        //output
        System.out.println( binarySearch(array,k));
        s1.close();
    }
}
